Implement name inference for addmv, addmv_, mv

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25181 [WIP][testing] Turn on BUILD_NAMEDTENSOR=1
* #25178 Include the correct header for make_unique in named tensor headers
* #25177 Implement name inference for torch.matmul
* #25123 Implement name inference for torch.bmm
* #24474 Implement name inference for torch.dot
* **#24471 Implement name inference for addmv, addmv_, mv**
* #24469 Implement name inference for expand
* #24306 Implement name inference for mm, addmm, addmm_
* #24921 Fix binary op name inference between unnamed and named tensors.

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

Differential Revision: [D16915805](https://our.internmc.facebook.com/intern/diff/D16915805)

zou3519 committed:
Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -248,17 +248,63 @@ void propagate_names(TensorImpl* result, TensorImpl* src) {
+    return;
   }
-  return mm_outnames;
+  auto mv_outnames = compute_dot_product_outnames(
```
this function has a lot of parameters, which can make it harder to read. 3 of them are passed as hard-coded 1, 2, 0 in all uses in this review. are there cases where they will be different or can we simplify this code?

nairbv replied:
I found other uses in another review/stack ... but still seems like there might be some simplification that can be made here. e.g. is `other_dotted_dim` always going to be 0? is other_ndim just the number of dims in `other`?

zou3519 replied:
> e.g. is other_dotted_dim always going to be 0

Probably.

> other_ndim just the number of dims in other

Only if `other` isn't nullopt. If `other` is nullopt then we don't know what's going on.

zou3519 replied:
I've changed it so that there are only 4 parameters instead of 6. I don't want to hardcode `other_dotted_dim` to 0, but I can do that if you think it would be better.

nairbv replied:
It's still not clear to me why not part of the implementation, and I think that could be confusing to anyone else who might want to call the function.

If it appears that it should always be zero, but it's a function parameter, personally I'd assume I miss-understood the meaning of the parameter. I would probably spend a lot of time trying to figure out what the other valid values would be/why before passing a zero. If there's a reason to keep it, maybe a comment would help explain why it's there or when something other than zero would make sense.

zou3519 replied:
For something like einsum, we'd use a dimension other than zero, however the function in its current state isn't sufficient for einsum. When one is doing tensor contraction (a generalized matrix multiply), one can pick one dimension from each input tensor to contract together. That dimension just happens to always be the last dimension of the first tensor and the zeroth dimension of the second tensor, but it isn't in the case of `einsum`.

zou3519 replied:
I'll go toss what I wrote in that comment into a comment in the code

nairbv replied:
that makes sense, I think an explanation in comments would be fine then

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 commented:
@nairbv this should be ready for another review.

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head

zou3519 committed:
Update on "Implement name inference for addmv, addmv_, mv"

Implement name inference for addmv, addmv_, mv

mv(Tensor[M, N], Tensor[O]) ignores the names of N and O and returns a
tensor with names [M].

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24471 gh/zou3519/107/head
